pkg/download/buffer.go

```diff
@@ -65,19 +67,34 @@ type firstReqResult struct {
 	err      error
 }
 
+func readBody(reader *multireader.BufferedReader, resp *http.Response) error {
+	expectedBytes := resp.ContentLength
+	_ = reader.SetSize(expectedBytes)
+	n, err := reader.ReadFrom(resp.Body)
+	if errors.Is(err, io.EOF) {
+		reader.Err = fmt.Errorf("error reading response for %s: %w", resp.Request.URL.String(), err)
+		return reader.Err
+	}
+	if n != expectedBytes {
+		reader.Err = fmt.Errorf("downloaded %d bytes instead of %d for %s", n, expectedBytes, resp.Request.URL.String())
+		return reader.Err
+	}
+	return nil
+}
+
 func (m *BufferMode) Fetch(ctx context.Context, url string) (io.Reader, int64, error) {
 	logger := logging.GetLogger()
 
-	br := newBufferedReader(m.minChunkSize())
+	br := multireader.NewBufferedReader(m.minChunkSize())
 
 	firstReqResultCh := make(chan firstReqResult)
 	m.queue.submit(func() {
 		m.sem.Go(func() error {
 			defer close(firstReqResultCh)
-			defer br.done()
+			defer br.Done()
 			firstChunkResp, err := m.DoRequest(ctx, 0, m.minChunkSize()-1, url)
 			if err != nil {
-				br.err = err
+				br.Err = err
```

Comment:
I'm confused by this. Are you using the `BufferedReader`'s `Err` field as an indirect signalling mechanism allowing `readBody` to return that error? If so, that seems... wrong. Why don't we just return an error here?